plugins/destination/managed_writer.go

```diff
@@ -19,68 +19,50 @@ type worker struct {
 
 func (p *Plugin) worker(ctx context.Context, metrics *Metrics, table *schema.Table, ch <-chan schema.CQTypes, flush <-chan chan bool) {
 	resources := make([][]any, 0)
+	sizeBytes := 0
 	for {
 		select {
 		case r, ok := <-ch:
-			//nolint:revive
-			if ok {
-				resources = append(resources, schema.TransformWithTransformer(p.client, r))
-				if len(resources) == p.spec.BatchSize {
-					start := time.Now()
-					if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
-						p.logger.Err(err).Str("table", table.Name).Int("len", p.spec.BatchSize).Dur("duration", time.Since(start)).Msg("failed to write batch")
-						// we don't return as we need to continue until channel is closed otherwise there will be a deadlock
-						atomic.AddUint64(&metrics.Errors, uint64(p.spec.BatchSize))
-					} else {
-						p.logger.Info().Str("table", table.Name).Int("len", p.spec.BatchSize).Dur("duration", time.Since(start)).Msg("batch written successfully")
-						atomic.AddUint64(&metrics.Writes, uint64(p.spec.BatchSize))
-					}
-					resources = make([][]any, 0)
-				}
-			} else {
+			if !ok {
 				if len(resources) > 0 {
-					start := time.Now()
-					if err := p.client.WriteTableBatch(ctx, table, resources); err != nil {
-						p.logger.Err(err).Str("table", table.Name).Int("len", len(resources)).Dur("duration", time.Since(start)).Msg("failed to write last batch")
-						atomic.AddUint64(&metrics.Errors, uint64(len(resources)))
-					} else {
-						p.logger.Info().Str("table", table.Name).Int("len", len(resources)).Dur("duration", time.Since(start)).Msg("last batch written successfully")
-						atomic.AddUint64(&metrics.Writes, uint64(len(resources)))
-					}
+					p.flush(ctx, metrics, table, resources)
 				}
 				return
 			}
+			if len(resources) == p.spec.BatchSize || sizeBytes+r.Size() > p.spec.BatchSizeBytes {
```

Comment:
Using zero as the default value shortcut breaks my heart every time. Maybe allow negative values (erm, `-1`) to mean "unlimited"?

Reply:
Sorry, not sure I understood the comment... Was it meant to be a comment on this line? If the batch size limit is 0, the SetDefaults call will ensure it's updated to the default batch size. So there will always be some limit, but you can of course set it to 1PB if you don't want it to affect anything in practice